### PR TITLE
Introduce post-processing pipeline

### PR DESCRIPTION
**Pull request by dictoon**

### CODE REVIEW

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 1]
```diff
@@ -297,17 +299,27 @@ struct MasterRenderer::Impl
...
+            // Don't proceed further if rendering failed.
+            if (result.m_status != RenderingResult::Succeeded)
+                return result;
 
-            return result;
+            // Post-process.
+            stopwatch.start();
+            postprocess(result);
+            stopwatch.measure();
+            result.m_post_processing_time = stopwatch.get_seconds();
```
Comment (oktomus) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: Why not storing it in `render_info` like `render_time` ? 
`render_info.insert("render_time", stopwatch.get_seconds());`

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 2]
```diff
@@ -549,37 +556,74 @@ struct MasterRenderer::Impl
...
+        // Detect post-processing stages with equal order.
+        size_t previous_stage_index = 0;
+        int previous_order = ordered_stages[0]->get_order();
+        for (size_t i = 1, e = ordered_stages.size(); i < e; ++i)
+        {
+            const int order = ordered_stages[i]->get_order();
+            if (order == previous_order)
+            {
+                RENDERER_LOG_WARNING(
+                    "post-processing stages \"%s\" and \"%s\" have equal order (%d); results will be unpredictable.",
```
Comment (oktomus) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: Does this shows up correctly on one line ?
Maybe something like : 
```
"following post-processing stages have equal order (%d):
- \"%s\" \n
- \"%s\" \n;
results will be unpredictable."
```

File: src/appleseed/renderer/modeling/environmentshader/edfenvironmentshader.cpp [hunk 3]
```diff
@@ -97,15 +98,17 @@ namespace
...
             // Cache the bound environment EDF.
             m_env_edf = dynamic_cast<const EnvironmentEDF*>(m_inputs.get_entity("environment_edf"));
 
             // Abort if no environment EDF was found.
             if (m_env_edf == nullptr)
             {
                 RENDERER_LOG_ERROR(
-                    "while preparing environment shader \"%s\": "
-                    "cannot find environment edf \"%s\".",
+                    "%scannot find environment edf \"%s\".",
```
Comment (oktomus) [inline src/appleseed/renderer/modeling/environmentshader/edfenvironmentshader.cpp]: Missing space

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 4]
```diff
@@ -100,26 +96,28 @@ UniqueID Frame::get_class_uid()
...
-    bool                    m_render_stamp_enabled;
-    string                  m_render_stamp_format;
-    DenoisingMode           m_denoising_mode;
-    bool                    m_save_extra_aovs;
+    size_t                          m_frame_width;
+    size_t                          m_frame_height;
+    size_t                          m_tile_width;
+    size_t                          m_tile_height;
+    string                          m_filter_name;
+    float                           m_filter_radius;
```
Comment (oktomus) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: Align name

File: src/appleseed/renderer/modeling/postprocessingstage/colormappostprocessingstage.cpp [hunk 5]
```diff
@@ -0,0 +1,1516 @@
...
+    // Perceptually uniform sequential color maps from Matplotlib, released under CC0 license.
+    //
+    // References:
+    //
+    //   http://bids.github.io/colormap/
+    //   https://github.com/BIDS/colormap/blob/master/colormaps.py
+    //   http://medvis.org/2016/02/23/better-than-the-rainbow-the-matplotlib-alternative-colormaps/
+    //
+
+    static const float MagmaColorMap[] =
```
Comment (oktomus) [inline src/appleseed/renderer/modeling/postprocessingstage/colormappostprocessingstage.cpp]: Maybe add these in another file ?

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp (hunk 1 above)
Comment (dictoon) [reply src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: I can but it won't be used at the moment (the only use of render info is the render stamp post-processing stage, which is executed before the total post-processing time is known).

File: src/appleseed/renderer/modeling/environmentshader/edfenvironmentshader.cpp (hunk 3 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/environmentshader/edfenvironmentshader.cpp]: On purpose. The first `%s` is either an empty string or a prefix such as `while defining entity X: ` (with a trailing space).

File: src/appleseed/renderer/modeling/frame/frame.cpp (hunk 4 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Everything is aligned, as far as I can see...

File: src/appleseed/renderer/modeling/postprocessingstage/colormappostprocessingstage.cpp (hunk 5 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/postprocessingstage/colormappostprocessingstage.cpp]: Done.

File: src/appleseed/renderer/modeling/frame/frame.cpp (hunk 4 above)
Comment (oktomus) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Weird..

Comment (dictoon) [reply src/appleseed/renderer/modeling/frame/frame.cpp]: Your browser is using a proportional font, for some reason...

**Pull request merged**